### ISSUE TITLE
[TS 4.5-dev] Regression: Argument of type is not assignable to parameter of type

### ISSUE THREAD

**Issue by eamodio** [microsoft/TypeScript#46643]
# Bug Report

Getting lots of issues with a template string inference pattern I use in GitLens for creating strongly typed versions of string-based settings. And in TS 4.5-beta (and the currently nightly) I am getting `Argument of type '* is not assignable to parameter of type` errors with that pattern

### 🕗 Version & Regression Information

- This is a crash
- This changed between versions `4.4.4` and `4.5-beta` and still present in `4.5.0-dev.202111101`
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________
- I was unable to test this on prior versions because _______

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.5.0-beta#code/C4TwDgpgBAygrgIwAoENgAsA8AVANFAaQhCggA9gIA7AEwGcoBrYgewDMpsA+KAXkOKkK1elDrAATgEsqAcwCwAKACQAfk4BtIiAC6QyrQYAlCAGMWEmpnHS5+FFRBclytS9fKAPlAAGAEgBvbQBfADpA+GQ0LGwtYh18AFEyUwAbOBoITGYQdk1tBKZWDgcQDR0uHgAyMUkZWWCfd1dvfyDiMMDktIysnLzYgvx+ksdy6trbBqaVZQAuKCoIADcICRcFpdWJAG4lJVBIKFQMHB5+SJOY4eLOHm8Rzn0RBht6qAfb7Ch1S+icG65DjcD5FIFPBaPbB7RQHcDQK4ANRQ6SyeGOz0Mx3+3HOGPIBlEbRkbDWAhAnQCJLJJnEjRc6m0mNEUPc6lpwGZDCuODiumcszU2IwyNRvKGUA5Ao8mxWa3csu2G3xwixrJU6liSB0yq2axh+0U5DAFk5MkoEjYKFM0AAwiwqGwpLIoAEXGwWCwFm7ZggUBJvc0-QAvBYIT2pCAOGGuYIxuNKYKGw52h1O2RXPjCrD2x3OrgwlNQXPppEouBZJBc4tp51XPFlsUl534JAFw3G01QNIoOgMZuyOASNBSB2ulyyCDAAAUAEoFgOY5PgDhqwP69P3HQzMBR1QFnh3JkrXBUsBRRXVAvaxnohe0VxcC55zW87eReWH0up6uCS9X6W0RcJuszbqYu4OlenBPrMx4oKe56flB653p+ZwwcoL4DqCKEfmKII+q4EhTkOVBQHAtAQE6Sw0FAvZ0Y48aGkmsKKOYVDiN2N5DiOY78EsADuAHOjxEFUHOMLsZxyyflm7HpqJe6hMu04AOQeiwoR+hIWkoMGqmzjsUAAPTGVAmAALQWVAACCEiDgAttQnJ5EWqkAEQabpOkhu5qlQFIDBUCwnK9nQzpUCgCCRlAwAsFAYD+igTkWlArnwlAHkaX5AVkdgMBQAALKEACsFkIFOKBKEAA)

### 💻 Code

```ts
type SubPath<T, Key extends keyof T> = Key extends string
	? T[Key] extends Record<string, any>
		?
				| `${Key}.${SubPath<T[Key], Exclude<keyof T[Key], keyof any[]>> & string}`
				| `${Key}.${Exclude<keyof T[Key], keyof any[]> & string}`
		: never
	: never;

type Path<T> = SubPath<T, keyof T> | keyof T extends string | keyof T ? SubPath<T, keyof T> | keyof T : keyof T;

type PathValue<T, P extends Path<T>> = P extends `${infer Key}.${infer Rest}`
	? Key extends keyof T
		? Rest extends Path<T[Key]>
			? PathValue<T[Key], Rest>
			: never
		: never
	: P extends keyof T
	? T[P]
	: never;


export interface Config {
	foo: {
		bar: {
			baz: boolean;
		};
	};
}

type ConfigPath = Path<Config>;
type ConfigPathValue<P extends ConfigPath> = PathValue<Config, P>;

export class Configuration {
	get(): Config;
	get<T extends ConfigPath>(
		section: T,
		defaultValue?: ConfigPathValue<T>,
	): ConfigPathValue<T>;
	get<T extends ConfigPath>(
		section?: T,
		defaultValue?: ConfigPathValue<T>,
	): Config | ConfigPathValue<T> {
		return undefined as any;
	}

}

const configuration = new Configuration();
const value = configuration.get('foo.bar.baz'); // <-- Argument of type '"foo.bar.baz"' is not assignable to parameter of type '"foo"' in TS 4.5-beta

```

<details><summary><b>Output</b></summary>

```ts
export class Configuration {
    get(section, defaultValue) {
        return undefined;
    }
}
const configuration = new Configuration();
const value = configuration.get('foo.bar.baz'); // <-- Argument of type '"foo.bar.baz"' is not assignable to parameter of type '"foo"' in TS 4.5-beta

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "strict": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "target": "ES2017",
    "jsx": "react",
    "module": "ESNext",
    "moduleResolution": "node"
  }
}
```

</details>
            
### 🙁 Actual behavior

`Argument of type '"foo.bar.baz"' is not assignable to parameter of type`

### 🙂 Expected behavior

No errors like in TS 4.4.4 and prior

**Comment by andrewbranch** [comment]
Can you explain your `Path<T>` type? When would the top-level condition not be true? Added parens for clarity:

> `(SubPath<T, keyof T> | keyof T) extends (string | keyof T)`

To me, it looks like `keyof T` can be factored out of each side of the union, so you’re left with asking if `SubPath<T, keyof T>` is assignable to `string`. Every possible instantiation of `SubPath` is assignable to `string`, so I’m missing what this condition is supposed to be checking for.

Bizarrely, a workaround that would seem not to change the semantics of these types at all is to extract `SubPath<T, keyof T> | keyof T` to [its own type alias](https://www.typescriptlang.org/play?ts=4.5.0-beta#code/C4TwDgpgBAygrgIwAoENgAsA8AVANFAaQhCggA9gIA7AEwGcoBrYgewDMpsA+KAXkOKkK1elDrAATgEsqAcwCwAKACQAfk4BtIiAC6QyrQYAlCAGMWEmpnHS5+FFRBclytS9fKAPlAAGAEgBvbQBfADpA+GQ0LGwtYh18AFEyUwAbOBoITGYQdk1tBKZWDgcQDR0uHgAyMUkZWWCfd1dvfyDiMMDktIysnLzYgvx+ksdy6trbBqaVZQAuKCoIADcICRcFpdWJAG4lJVBIWERUDAB5CW12bBwefkjTmOHizh5vEc49xUPoR9u+Y5Rc6XYo3bj6EQMGz1KDvF7YKDqB7RHDPXIccFw9GcKALD7YL4HcC-aIANRQ6SyeCgSAhhhpKO4dxpdNEbRkbDWAhAnQCHK5JnEjRc6m0rIY+Pc6kFwHFDIwODiumcszU8vQ5MpiqGUBlKo8mxWa3chu2GxZ5AMoklKnUsSQOnNWzWhMUSnIYAsspklAkbBQpmgAGEWFQ2FJZFAAi42CwWAto7MECgJAnmsmAF4LBBx1IQBxfVzBQvFpTBfbfYlQENhiOPAF-Gvh2RcL4-auh5uPTVwLK0y2Qju12SPZndim9zBNiP4JCtiser1QNIoOgMaeyOASNBSUNRlyyCDAAAUAEoFhvC4fgDg5RvR8f3HQzMBd1QFnh3Jl-XBUsAexAqgXp2dZkhOVJcLgLjnkOXZgVq3BXket4DvS97RFwj6zM+pivqGQGcFBszfigv7-uBBHoRgAG3ERygwRusKwaB1Hgf8iauBIR5blQUBwLQEDhksNBQKuomOCWFblm6ijmFQ4jLiBm7bnhvH8EsADuzHKTuoZnl8ckKcs4EAnJzZbrpVChNex4AOSxiwoTJhITkoBmtmnjsUAAPTeVAmAALQBVAACCEibgAttQsp5O2tkAEQOa5LmZvFtlQFIDBUCwsqrnQEZUCgCB5lAwAsFAYApigUW+lAsVVglDlpRlvHYDAUAACyhAArAFCBHigShAA). That definitely seems like a bug, but it’s hard for me to figure out how bad of a break this is because I can’t figure out the reason for the type at all. It seems to me like you could just as well define

```ts
type Path<T> = SubPath<T, keyof T> | keyof T;
```

and this example works as expected.

**Comment by ahejlsberg** [comment]
This appears to be another manifestation of #46645. The issue disappears from the `main` branch after [this](https://github.com/microsoft/TypeScript/commit/56f81075f9802c85365ab24046e7eb947cd7800a) commit, so it should also be fixed in the `release-4.5` branch after the cherry pick.

**Comment by eamodio** [comment]
@andrewbranch I think I thought I needed that extra complexity on `SubPath` to allow for deeper nesting, but it looks like `type Path<T> = SubPath<T, keyof T> | keyof T;` as you suggested works just fine. Thanks!
